client/rest/src/main/java/org/elasticsearch/client/RestClient.java

```diff
@@ -372,157 +473,61 @@ private static URI buildUri(String path, Map<String, String> params) {
-            return this;
-        }
+    private static class SyncResponseListener implements ResponseListener {
+        final CountDownLatch latch = new CountDownLatch(1);
+        volatile Response response;
+        volatile Exception exception;
 
-        /**
-         * Sets the {@link RequestConfigCallback} to be used to customize http client configuration
-         */
-        public Builder setRequestConfigCallback(RequestConfigCallback requestConfigCallback) {
-            Objects.requireNonNull(requestConfigCallback, "requestConfigCallback must not be null");
-            this.requestConfigCallback = requestConfigCallback;
-            return this;
+        @Override
+        public void onSuccess(Response response) {
+            this.response = response;
+            latch.countDown();
         }
 
-        /**
-         * Creates a new {@link RestClient} based on the provided configuration.
-         */
-        public RestClient build() {
-            if (failureListener == null) {
-                failureListener = new FailureListener();
-            }
-            CloseableHttpClient httpClient = createHttpClient();
-            return new RestClient(httpClient, maxRetryTimeout, defaultHeaders, hosts, failureListener);
+        @Override
+        public void onFailure(Exception exception) {
+            this.exception = exception;
+            latch.countDown();
         }
 
-        private CloseableHttpClient createHttpClient() {
-            //default timeouts are all infinite
-            RequestConfig.Builder requestConfigBuilder = RequestConfig.custom().setConnectTimeout(DEFAULT_CONNECT_TIMEOUT_MILLIS)
-                    .setSocketTimeout(DEFAULT_SOCKET_TIMEOUT_MILLIS)
-                    .setConnectionRequestTimeout(DEFAULT_CONNECTION_REQUEST_TIMEOUT_MILLIS);
-
-            if (requestConfigCallback != null) {
-                requestConfigCallback.customizeRequestConfig(requestConfigBuilder);
-            }
-            RequestConfig requestConfig = requestConfigBuilder.build();
-
-            PoolingHttpClientConnectionManager connectionManager = new PoolingHttpClientConnectionManager();
-            //default settings may be too constraining
-            connectionManager.setDefaultMaxPerRoute(10);
-            connectionManager.setMaxTotal(30);
-
-            HttpClientBuilder httpClientBuilder = HttpClientBuilder.create().setConnectionManager(connectionManager)
-                    .setDefaultRequestConfig(requestConfig);
-
-            if (httpClientConfigCallback != null) {
-                httpClientConfigCallback.customizeHttpClient(httpClientBuilder);
+        Response get() throws Exception {
+            latch.await();
+            if (response != null) {
+                assert exception == null;
+                return response;
             }
-            return httpClientBuilder.build();
+            assert exception != null;
+            throw exception;
```

Comment:
I imagine this'll get funky stack traces. Maybe wrap it? If you wrap it in something that extends RuntimeException client's won't have to catch it which is *mostly* a good thing. It is certainly the way that frameworks like Spring like to work.